lib/tpm2_alg_util.c

```diff
@@ -0,0 +1,143 @@
+#include <string.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include <sapi/tpm20.h>
+
+#include "tpm2_util.h"
+#include "tpm2_alg_util.h"
+
+typedef struct alg_pair alg_pair;
+struct alg_pair {
+    const char *name;
+    TPM_ALG_ID id;
+};
+
+/*
+ * The TSS has a bug where it was missing algs 0x27 trough 0x29.
+ * see: https://github.com/01org/TPM2.0-TSS/issues/476
+ * per https://trustedcomputinggroup.org/wp-content/uploads/TCG_Algorithm_Registry_Rev_1.24.pdf
+ */
+#ifndef ALG_SHA3_256_VALUE
+#define ALG_SHA3_256_VALUE 0x27
+#endif
+#ifndef ALG_SHA3_384_VALUE
+#define ALG_SHA3_384_VALUE 0x28
+#endif
+#ifndef ALG_SHA3_512_VALUE
+#define ALG_SHA3_512_VALUE 0x29
+#endif
+
+void tpm2_alg_util_for_each_alg(tpm2_alg_util_alg_iteraror iterator, void *userdata) {
+
```

Comment:
Well it is iterating over the contents, what you choose to do with it is arbitrary. I was torn between just making the data extern vs encapsulating it like this. The encapsulation here insn't buying us much except for avoiding the writing of a for loop for what that's worth.